bitcoin/src/psbt/serialize.rs

```diff
@@ -42,40 +43,57 @@ impl Psbt {
     /// Serialize as raw binary data
     pub fn serialize(&self) -> Vec<u8> {
         let mut buf: Vec<u8> = Vec::new();
+        self.serialize_to_writer(&mut buf).expect("Writing to Vec can't fail");
+        buf
+    }
 
-        //  <magic>
-        buf.extend_from_slice(b"psbt");
+    /// Serialize the PSBT into a writer.
+    pub fn serialize_to_writer(&self, w: &mut impl Write) -> io::Result<usize> {
+        let mut written_len = 0;
 
-        buf.push(0xff_u8);
+        fn write_all(w: &mut impl Write, data: &[u8]) -> io::Result<usize> {
+            w.write_all(data).map(|_| data.len())
+        }
 
-        buf.extend(self.serialize_map());
+        // magic
+        written_len += write_all(w, b"psbt")?;
+        // separator
+        written_len += write_all(w, &[0xff])?;
+
+        written_len += write_all(w, &self.serialize_map())?;
 
         for i in &self.inputs {
-            buf.extend(i.serialize_map());
+            written_len += write_all(w, &i.serialize_map())?;
         }
 
         for i in &self.outputs {
-            buf.extend(i.serialize_map());
+            written_len += write_all(w, &i.serialize_map())?;
         }
 
-        buf
+        Ok(written_len)
     }
 
     /// Deserialize a value from raw binary data.
-    pub fn deserialize(bytes: &[u8]) -> Result<Self, Error> {
+    pub fn deserialize(mut bytes: &[u8]) -> Result<Self, Error> {
+        Self::deserialize_from_reader(&mut bytes)
+    }
+
+    /// Deserialize a value from raw binary data read from a `BufRead` object.
+    pub fn deserialize_from_reader<R: io::BufRead>(r: &mut R) -> Result<Self, Error> {
         const MAGIC_BYTES: &[u8] = b"psbt";
-        if bytes.get(0..MAGIC_BYTES.len()) != Some(MAGIC_BYTES) {
+
+        let magic: [u8; 4] = Decodable::consensus_decode(r)?;
+        if magic != MAGIC_BYTES {
             return Err(Error::InvalidMagic);
         }
 
         const PSBT_SERPARATOR: u8 = 0xff_u8;
-        if bytes.get(MAGIC_BYTES.len()) != Some(&PSBT_SERPARATOR) {
+        let separator: u8 = Decodable::consensus_decode(r)?;
```

Comment:
Might be slightly better to just read 5 bytes at once and compare them in one go.

Reply:
Can do that, although I think this version is slightly more readable.

Reply:
Given that `error[E0277]: the trait bound `[u8; 5]: Decodable` is not satisfied`, I'd have to use `read_exact` like so:

```patch
diff --git a/bitcoin/src/psbt/serialize.rs b/bitcoin/src/psbt/serialize.rs
index 9fa465d1..d201ff45 100644
--- a/bitcoin/src/psbt/serialize.rs
+++ b/bitcoin/src/psbt/serialize.rs
@@ -81,15 +81,16 @@ pub fn deserialize(mut bytes: &[u8]) -> Result<Self, Error> {
     /// Deserialize a value from raw binary data read from a `BufRead` object.
     pub fn deserialize_from_reader<R: io::BufRead>(r: &mut R) -> Result<Self, Error> {
         const MAGIC_BYTES: &[u8] = b"psbt";
+        const PSBT_SERPARATOR: u8 = 0xff_u8;
+
+        let mut magic_and_separator: [u8; 5] = [0; 5];
+        r.read_exact(&mut magic_and_separator)?;
 
-        let magic: [u8; 4] = Decodable::consensus_decode(r)?;
-        if magic != MAGIC_BYTES {
+        if &magic_and_separator[..4] != MAGIC_BYTES {
             return Err(Error::InvalidMagic);
         }
 
-        const PSBT_SERPARATOR: u8 = 0xff_u8;
-        let separator: u8 = Decodable::consensus_decode(r)?;
-        if separator != PSBT_SERPARATOR {
+        if magic_and_separator[4] != PSBT_SERPARATOR {
             return Err(Error::InvalidSeparator);
         }
 
```

I don't think that's an improvement overall (especially since you already force buffered readers, which should make the benefit of not doing a one-byte read negligible).